ludwig/combiners/combiners.py

```diff
@@ -41,23 +42,65 @@
 logger = logging.getLogger(__name__)
 
 
-class ConcatCombiner(LudwigModule):
+# super class to house common properties
+class CombinerClass(LudwigModule):
+    @property
+    def effective_input_shape(self) -> torch.Size:
+        # computes the effective shape of the input tensor after combining
+        # all the encoder outputs
+        # determine sequence size by finding the first sequence tensor
+        # assume all the sequences are of the same size, if not true
+        # this will be caught during processing
+        seq_size = None
+        for k in self.input_features:
+            # dim-2 output_shape implies a sequence [seq_size, hidden]
+            if len(self.input_features[k].output_shape) == 2:
+                seq_size = self.input_features[k].output_shape[0]
+                break
+
+        # collect the size of the last dimension for all input feature
+        # encoder outputs
+        shapes = [self.input_features[k].output_shape[-1] for k in
+                  self.input_features]  # output shape not input shape
+        return torch.Size([seq_size, sum(shapes)])
+
+    @property
+    def input_shape(self) -> dict:
+        # input to combiner is a dictionary of the input features encoder
+        # outputs, this property returns dictionary of output shapes for each
+        # input feature's encoder output shapes.
+        return {k: self.input_features[k].output_shape
+                for k in self.input_features}
+
+    @property
+    @lru_cache(maxsize=1)
+    def output_shape(self) -> torch.Size:
+        psuedo_input = {k: {'encoder_output': torch.rand(2,
+                                                         *self.input_features[
+                                                             k].output_shape,
+                                                         dtype=self.input_dtype)}
+                        for k in self.input_features}
+        output_tensor = self.forward(psuedo_input)
+        return output_tensor['combiner_output'].size()[1:]
+
+
+class ConcatCombiner(CombinerClass):
     def __init__(
             self,
-            input_features: dict = None,
+            input_features: typing.Dict = None,
```

Comment:
why not importing it?

Reply:
@ShreyaR  suggested this particular  [review comment in another PR](https://github.com/ludwig-ai/ludwig/pull/1332#discussion_r718931809).  As I understand it, this particular form will allow us to specify data types for key and values in the future if needed.

Reply:
Right, I should have been cleared. I meant why not importing Dict from typing instead of referring to it as typing.Dict ?

Reply:
done.